src/layer.cpp

```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
+/// @yield [Events::Touch] the given block will be yield when a touch end
+///   event is received.
+/// @return [Scene] the receiver.
+
+static VALUE
+scene_on_touch_end(VALUE rcv, SEL sel)
+{
+    VALUE block = rb_current_block();
+    if (block == Qnil) {
+    rb_raise(rb_eArgError, "block not given");
+    }
+    block = rb_retain(block); // FIXME need release...
+
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
+        scene->touch_listener = cocos2d::EventListenerTouchOneByOne::create();
+    } else {
+        scene_remove_listener(rcv, scene->touch_listener);
+    }
+    scene->touch_listener->onTouchEnded = [block](cocos2d::Touch *touch,
+        cocos2d::Event *event) -> bool {
+    VALUE touch_obj = rb_class_wrap_new((void *)touch,
+        rb_cTouch);
+    return RTEST(rb_block_call(block, 1, &touch_obj));
+    };
+
+    return scene_add_listener(rcv, scene->touch_listener);
+}
+
+/// @method #on_touch_move
+/// Starts listening for touch move events on the receiver.
+/// @yield [Events::Touch] the given block will be yield when a touch move
+///   event is received.
+/// @return [Scene] the receiver.
+
+static VALUE
+scene_on_touch_move(VALUE rcv, SEL sel)
+{
+    VALUE block = rb_current_block();
+    if (block == Qnil) {
+    rb_raise(rb_eArgError, "block not given");
+    }
+    block = rb_retain(block); // FIXME need release...
+
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
+        scene->touch_listener = cocos2d::EventListenerTouchOneByOne::create();
+    } else {
+        scene_remove_listener(rcv, scene->touch_listener);
+    }
+    scene->touch_listener->onTouchMoved = [block](cocos2d::Touch *touch,
+        cocos2d::Event *event) -> bool {
+    VALUE touch_obj = rb_class_wrap_new((void *)touch,
+        rb_cTouch);
+    return RTEST(rb_block_call(block, 1, &touch_obj));
+    };
+
+    return scene_add_listener(rcv, scene->touch_listener);
 }
 
+/// @method #on_touch_cancel
+/// Starts listening for touch cancel events on the receiver.
+/// @yield [Events::Touch] the given block will be yield when a touch cancel
+///   event is received.
+/// @return [Scene] the receiver.
+
+static VALUE
+scene_on_touch_cancel(VALUE rcv, SEL sel)
+{
+    VALUE block = rb_current_block();
+    if (block == Qnil) {
+    rb_raise(rb_eArgError, "block not given");
+    }
+    block = rb_retain(block); // FIXME need release...
+
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
+        scene->touch_listener = cocos2d::EventListenerTouchOneByOne::create();
+    } else {
+        scene_remove_listener(rcv, scene->touch_listener);
+    }
+    scene->touch_listener->onTouchCancelled = [block](cocos2d::Touch *touch,
+        cocos2d::Event *event) -> bool {
+    VALUE touch_obj = rb_class_wrap_new((void *)touch,
+        rb_cTouch);
+    return RTEST(rb_block_call(block, 1, &touch_obj));
+    };
+
+    return scene_add_listener(rcv, scene->touch_listener);
+}
```

Comment:
I think we should refactor the content of these functions (`on_touch_{begin,end,move.cancel}`) before merging this patch, as they are almost identical. Creating a new function with the common code and calling it from the other functions.